python_modules/dagster/dagster/_core/instance/storage/storage_domain.py

```diff
@@ -0,0 +1,208 @@
+"""Storage domain implementation - extracted from DagsterInstance."""
+
+from collections.abc import Mapping, Sequence
+from typing import TYPE_CHECKING, Any, Callable, Optional
+
+from dagster._core.types.pagination import PaginatedResults
+
+# Type alias for print function to avoid circular imports
+PrintFn = Callable[[Any], None]
+
+if TYPE_CHECKING:
+    from dagster._core.events import AssetKey, DagsterEventType
+    from dagster._core.instance import DagsterInstance
+
+
+class StorageDomain:
+    """Domain object encapsulating storage-related operations.
+
+    This class holds a reference to a DagsterInstance and provides methods
+    for dynamic partitions, storage optimization, and directory management.
```

Comment:
grouping dynamic partitions and directory management together here feels slightly odd, but maybe splitting them apart gets overly-granular